File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        subset = []
        index = 0
        self.backtrack(res, subset, index, nums, len(nums))

        return res

    def backtrack(self, res, subset, index, nums, length):
        if len(subset) == length:
            res.append(subset.copy());
            return

        subset.append(nums[index])
        popped = nums.pop(index)
        self.backtrack(res, subset, 0, nums, length)

        subset.pop(len(subset)-1)
        nums.insert(index, popped)
        if index <= len(nums)-2:
            self.backtrack(res, subset, index+1, nums, length)

File: Data Structures & Algorithms/longest-palindromic-substring/submission-2.java
class Solution {
    public String longestPalindrome(String s) {
        String res = "";
        char[] str = s.toCharArray();
        for(int i = 0;i < str.length;i++) {
            int l, r;
            l = r = i;
            while(l >= 0 && r < str.length && str[l] == str[r]) {
                if (res.length() < r-l+1) {
                    res = s.substring(l, r+1);
                }
                l--;
                r++;
            }   
            l = i;
            r = i+1;
            while(l >= 0 && r < str.length && str[l] == str[r]) {
                if (res.length() < r-l+1) {
                    res = s.substring(l, r+1);
                }
                l--;
                r++;
            }   
        }

        return res;
    }
}
